core/src/main/java/com/linecorp/armeria/client/Http2ClientConnectionHandler.java

```diff
@@ -110,7 +117,8 @@ public void channelActive(ChannelHandlerContext ctx) throws Exception {
 
     @Override
     protected boolean needsImmediateDisconnection() {
-        return clientFactory.isClosing() ||
+        return gracefulConnectionShutdownHandler.hasDrainStarted() ||
```

Comment:
In `Http2ServerConnectionHandler`, it uses `gracefulShutdownSupport.isShuttingDown()` to check whether it needs immediate disconnection like below. 
```java
@Override
protected boolean needsImmediateDisconnection() {
    return gracefulShutdownSupport.isShuttingDown() ||
           requestDecoder.goAwayHandler().receivedErrorGoAway() || keepAliveHandler().isClosing();
}
```

I couldn't find out why we need both `GracefulShutdownSupport` and `Http2GracefulConnectionShutdownHandler` are both needed 🤔. So instead, I've added `hasDrainStarted` method to check whether connection drain has started.

Reply:
Oh, why does it prevent sending the final `GOAWAY` frame? Even when `hasDrainStarted` is `true`, it seems that the code can still reach `goAway(ctx, connection().remote().lastStreamCreated()).addListener(future -> super.close(ctx, promise));`

Reply:
Ah, that's right. I misread the code. 😓 
However, it still closes all streams even if we shouldn't, let's revert it. 😉

Reply:
Thank you, now I finally get to understand a bit. (btw I've removed `hasDrainStarted`) 

=> We don't want the streams to be closed when draining starts. Because closing the streams right after draining starts doesn't seem to be appropriate for graceful disconnection.